lib/phoenix_component.ex

```diff
@@ -2907,8 +2913,35 @@ defmodule Phoenix.Component do
 
   slot.(:inner_block, [])
 
-  def dynamic_tag(%{name: name, rest: rest} = assigns) do
-    tag_name = to_string(name)
+  def dynamic_tag(%{rest: rest} = assigns) do
+    {tag_name, name} =
+      case assigns do
+        %{tag_name: tag_name, name: name} ->
+          {tag_name, name}
+
+        %{tag_name: tag_name} ->
+          {tag_name, nil}
+
+        %{name: name} ->
+          IO.warn("""
+          Passing the tag name to `Phoenix.Component.dynamic_tag/1` using the `name` attribute is deprecated.
+
+          Instead of:
+
+              <.dynamic_tag name="p" ...>
+
+          use `tag_name` instead:
+
+              <.dynamic_tag tag_name="p" ...>
+          """)
+
+          {name, nil}
```

Comment:
```suggestion
    {tag_name, rest} =
      case assigns do
        %{tag_name: tag_name, name: name} ->
          {tag_name, Map.put(rest, :name, name)}

        %{tag_name: tag_name} ->
          {tag_name, rest}

        %{name: name} ->
          IO.warn("""
          Passing the tag name to `Phoenix.Component.dynamic_tag/1` using the `name` attribute is deprecated.

          Instead of:

              <.dynamic_tag name="p" ...>

          use `tag_name` instead:

              <.dynamic_tag tag_name="p" ...>
          """)

          {name, Map.delete(:rest, :name)}
```